File: janus/src/providers/openai.ts
import OpenAI from 'openai';
import { Provider, CompletionOptions } from './base.js';
import { ProviderConfig } from '../core/types.js';

export class OpenAIProvider implements Provider {
    public name = 'OpenAI';
    private client: OpenAI;
    private model: string;

    constructor(config: ProviderConfig) {
        this.model = config.model || 'gpt-4o';
        this.client = new OpenAI({
            apiKey: config.api_key || process.env.OPENAI_API_KEY,
        });
    }

    async generate(prompt: string, options?: CompletionOptions): Promise<string> {
        try {
            // GPT-5.x models use max_completion_tokens, older models use max_tokens
            const isGpt5 = this.model.startsWith('gpt-5') || this.model.startsWith('o1') || this.model.startsWith('o3');

            const requestParams: any = {
                messages: [
                    { role: 'system', content: options?.systemPrompt || 'You are a helpful assistant.' },
                    { role: 'user', content: prompt }
                ],
                model: this.model,
            };

            // GPT-5.x and reasoning models don't support temperature
            if (!isGpt5) {
                requestParams.temperature = options?.temperature ?? 0.7;
            }

            // Use appropriate token parameter based on model
            if (options?.maxTokens) {
                if (isGpt5) {
                    requestParams.max_completion_tokens = options.maxTokens;
                } else {
                    requestParams.max_tokens = options.maxTokens;
                }
            }

            const completion = await this.client.chat.completions.create(requestParams);

            return completion.choices[0].message.content || '';
        } catch (error) {
            console.error(`OpenAI Error:`, error);
            return `[Error] OpenAI failed to generate response: ${error}`;
        }
    }
}

File: python/janus/providers/openai.py
"""OpenAI provider implementation."""

import os

from openai import AsyncOpenAI

from ..core.types import ProviderConfig
from .base import CompletionOptions, Provider


class OpenAIProvider(Provider):
    """Provider for OpenAI's API."""

    def __init__(self, config: ProviderConfig, name: str = "OpenAI"):
        super().__init__(name)
        self.model = config.model or "gpt-5.2"
        self.client = AsyncOpenAI(
            api_key=config.api_key or os.environ.get("OPENAI_API_KEY"),
            base_url=config.base_url,
        )

    async def generate(self, prompt: str, options: CompletionOptions | None = None) -> str:
        """Generate a completion using OpenAI's API."""
        options = options or CompletionOptions()

        try:
            messages = []
            if options.system_prompt:
                messages.append({"role": "system", "content": options.system_prompt})
            messages.append({"role": "user", "content": prompt})

            # GPT-5.x and o1/o3 models use different parameters
            is_new_model = any(
                self.model.startswith(prefix) for prefix in ("gpt-5", "o1", "o3")
            )

            kwargs = {
                "model": self.model,
                "messages": messages,
            }

            if is_new_model:
                kwargs["max_completion_tokens"] = options.max_tokens
                # New models don't support temperature
            else:
                kwargs["max_tokens"] = options.max_tokens
                if options.temperature is not None:
                    kwargs["temperature"] = options.temperature

            response = await self.client.chat.completions.create(**kwargs)
            return response.choices[0].message.content or ""

        except Exception as e:
            return f"[Error] {self.name} failed to generate response: {e}"
